libraries/Microsoft.Bot.Builder.Dialogs/DialogManager.cs

```diff
@@ -186,92 +181,15 @@ public async Task<DialogManagerResult> OnTurnAsync(ITurnContext context, Cancell
-                dc.Services[service.Key] = service.Value;
-            }
-
-            // get the DialogStateManager configuration
-            var dialogStateManager = new DialogStateManager(dc, StateConfiguration);
-            await dialogStateManager.LoadAllScopesAsync(cancellationToken).ConfigureAwait(false);
-            dc.Context.TurnState.Add(dialogStateManager);
-
-            DialogTurnResult turnResult = null;
-
-            // Loop as long as we are getting valid OnError handled we should continue executing the actions for the turn.
-            //
-            // NOTE: We loop around this block because each pass through we either complete the turn and break out of the loop
-            // or we have had an exception AND there was an OnError action which captured the error.  We need to continue the 
-            // turn based on the actions the OnError handler introduced.
-            var endOfTurn = false;
-            while (!endOfTurn)
-            {
-                try
-                {
-                    if (context.TurnState.Get<IIdentity>(BotAdapter.BotIdentityKey) is ClaimsIdentity claimIdentity && SkillValidation.IsSkillClaim(claimIdentity.Claims))
-                    {
-                        // The bot is running as a skill.
-                        turnResult = await HandleSkillOnTurnAsync(dc, cancellationToken).ConfigureAwait(false);
-                    }
-                    else
-                    {
-                        // The bot is running as root bot.
-                        turnResult = await HandleBotOnTurnAsync(dc, cancellationToken).ConfigureAwait(false);
-                    }
-
-                    // turn successfully completed, break the loop
-                    endOfTurn = true;
-                }
-                catch (Exception err)
-                {
-                    // fire error event, bubbling from the leaf.
-                    var handled = await dc.EmitEventAsync(DialogEvents.Error, err, bubble: true, fromLeaf: true, cancellationToken: cancellationToken).ConfigureAwait(false);
-
-                    if (!handled)
-                    {
-                        // error was NOT handled, throw the exception and end the turn. (This will trigger the Adapter.OnError handler and end the entire dialog stack)
-                        throw;
-                    }
-                }
-            }
-
-            // save all state scopes to their respective botState locations.
-            await dialogStateManager.SaveAllChangesAsync(cancellationToken).ConfigureAwait(false);
+            // Call the common dialog "continue/begin" execution pattern shared with the classic RunAsync extension method
+            var turnResult = await DialogExtensions.InternalRunAsync(context, _rootDialogId, dc, cancellationToken).ConfigureAwait(false);
 
             // save BotState changes
             await botStateSet.SaveAllChangesAsync(dc.Context, false, cancellationToken).ConfigureAwait(false);
```

Comment:
is this needed?  since DialogExtensions.InternalRunAsync doeds this internally?

Reply:
Sorry @johnataylor, while looking this over i was confused between the DialogStateManager's SaveAllChangesAsync and the BotStateSet.SaveAllChangesAsync  :(